### PR TITLE
[ENH] OWMDS, OWLinearProjection: save selection in workflow

### PR DESCRIPTION
**Pull request by kernc**
##### Issue
Fixes https://github.com/biolab/orange3/issues/2182

##### Includes
- [X] Code changes
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by astaric** [comment]
Selections are not really context settings as they depend not only on the columns in the dataset but on the number of rows as well. Have you considered using schema only settings?

**Comment by kernc** [comment]
The problem I have with schema_only settings is that there's additional boilerplate associated with using them:
```py
    def closeContext(self):
        if self.current_context is not None:
            self.selection = None
        super().closeContext()
```
Short, but very intricate.

**Comment by janezd** [comment]
File -> Sample -> MDS. Select all points in MDS. Decrease the sample size in Sample. MDS crashes due to selection indices out of range. Linear projection probably as well.

A saved context setting is reused when "useful", e.g. if the setting includes a variable, the variable has to appear in the domain for the context to be useful. In this sense, the context here is not the domain but the sample size (e.g. the saved selection is "useful" if there is enough data for these indices). However, this context would be too vaguely defined, so checking the domain (`DomainContextHandler`) is OK.

One quick solution would be to discard the selection (after `openContext`) if `max(selection_indices) >= len(data)`. A slightly better solution is to have `DomainAndDataSizeContextHandler`, but it can still lead to weird selections. The safest solution would be `DataHashContextHandler`, where the has is computed over `X`, `Y` and `metas`. The latter would also make sense for some other widgets.

**Comment by kernc** [comment]
Right. How is this?

File: Orange/widgets/widget.py [hunk 1]
```diff
@@ -508,6 +508,14 @@ def closeContext(self):
         reinitializing the user interface (e.g. combo boxes) with the new
         data.
         """
+        # Clear schema-only settings when *closing* the context
+        # FIXME: Sadly, schema-only settings aren't cleared when non-contextual
```
Comment (astaric) [inline Orange/widgets/widget.py]: Actually, they are.

As non-context-aware widgets only load settings when a new instance of the widget is put on the canvas, this is achieved by not updating default values. See https://github.com/biolab/orange3/blob/master/Orange/widgets/settings.py#L508

File: Orange/widgets/widget.py [hunk 2]
```diff
@@ -508,6 +508,14 @@ def closeContext(self):
...
+        # SettingsHandler is used
+        if self.current_context is not None:
```
Comment (astaric) [inline Orange/widgets/widget.py]: With this change, https://github.com/biolab/orange3/blob/master/Orange/widgets/visualize/owscatterplot.py#L581 is now redundant.

File: Orange/widgets/widget.py (hunk 1 above)
Comment (kernc) [inline Orange/widgets/widget.py]: But what if the _same instance_ of widget but passed different data, as described [here](https://github.com/biolab/orange3/pull/2301/files#diff-2281ecd412ed7edd749aca4486d1f338R779)?

**Comment by astaric** [comment]
This looks good to me, rebase when possible so we can merge.

File: Orange/widgets/unsupervised/owmds.py [hunk 3]
```diff
@@ -103,7 +103,7 @@ class Outputs:
 
     settingsHandler = settings.DomainContextHandler()
 
-    max_iter = settings.Setting(300)
+    max_iter = settings.Setting(10)
```
Comment (lanzagar) [inline Orange/widgets/unsupervised/owmds.py]: What is this change doing in a commit "save selection in workflow" ?!?

Did you change this while testing and accidentally committed it?
Or decided to change some defaults and make it less obvious by not mentioning it at all and/or putting it in its own commit.

**Pull request merged**